core/src/main/java/org/apache/druid/data/input/impl/JsonReader.java

```diff
@@ -49,41 +55,99 @@
       boolean keepNullColumns
   )
   {
-    super(inputRowSchema, source);
+    this.inputRowSchema = inputRowSchema;
+    this.source = source;
     this.flattener = ObjectFlatteners.create(flattenSpec, new JSONFlattenerMaker(keepNullColumns));
     this.mapper = mapper;
   }
 
-  @Override
-  public List<InputRow> parseInputRows(String line) throws IOException, ParseException
-  {
-    final JsonNode document = mapper.readValue(line, JsonNode.class);
-    final Map<String, Object> flattened = flattener.flatten(document);
-    return Collections.singletonList(MapInputRowParser.parse(getInputRowSchema(), flattened));
-  }
 
   @Override
-  public Map<String, Object> toMap(String intermediateRow) throws IOException
+  public CloseableIterator<InputRow> read() throws IOException
   {
-    //noinspection unchecked
-    return mapper.readValue(intermediateRow, Map.class);
+    final MappingIterator<JsonNode> delegate = mapper.readValues(
+        new JsonFactory().createParser(this.source.open()),
```

Comment:
The `parser` created from `JsonFactory` is `Closeable` which should be closed when the below `CloseableIterator` is closed.